### PR TITLE
Oauth 4 okta

### PR DESCRIPTION
**Pull request by bretterer**
Adding Oauth for Okta!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -377,6 +394,32 @@ okta:
...
+    oktaDomain: "https://{yourOktaDomain}"
+    proxy:
+      port: null
+      host: null
+      username: null
+      password: null
+    authorizationMode: "PrivateKey"
+    clientId: "{yourClientId}"
+    Scopes: scope.1 scope.2
+    privateKey: |
```
Comment (aarongranick-okta) [inline README.md]: should this be lower-case "scopes" ?

File: okta/config.go [hunk 2]
```diff
@@ -41,8 +41,12 @@ type config struct {
 			RateLimit         struct {
 				MaxRetries int32 `yaml:"maxRetries" envconfig:"OKTA_CLIENT_RATE_LIMIT_MAX_RETRIES"`
 			} `yaml:"rateLimit"`
-			OrgUrl string `yaml:"orgUrl" envconfig:"OKTA_CLIENT_ORGURL"`
-			Token  string `yaml:"token" envconfig:"OKTA_CLIENT_TOKEN"`
+			OrgUrl            string   `yaml:"orgUrl" envconfig:"OKTA_CLIENT_ORGURL"`
+			Token             string   `yaml:"token" envconfig:"OKTA_CLIENT_TOKEN"`
+			AuthorizationMode string   `yaml:"authorizationMode" envconfig:"OKTA_CLIENT_AUTHORIZATIONMODE"`
+			ClientId          string   `yaml:"clientId" envconfig:"OKTA_CLIENT_CLIENTID"`
+			Scopes            []string `yaml:"scopes" envconfig:"OKTA_CLIENT_SCOPES"`
```
Comment (aarongranick-okta) [inline okta/config.go]: In the README scopes is shown in the yaml as a single string, separated by spaces. Is there any need for special handling for this to be an array here?

File: okta/requestExecutor.go [hunk 3]
```diff
@@ -74,7 +95,83 @@ func (re *RequestExecutor) NewRequest(method string, url string, body interface{
...
+	if re.config.Okta.Client.AuthorizationMode == "SSWS" {
+		req.Header.Add("Authorization", "SSWS "+re.config.Okta.Client.Token)
+	}
+
+	if re.config.Okta.Client.AuthorizationMode == "PrivateKey" {
+		if re.cache.Has("OKTA_ACCESS_TOKEN") {
+			token := re.cache.GetString("OKTA_ACCESS_TOKEN")
+			req.Header.Add("Authorization", "Bearer "+token)
+		} else {
+			priv := []byte(re.config.Okta.Client.PrivateKey)
```
Comment (aarongranick-okta) [inline okta/requestExecutor.go]: In the config, `PrivateKey` is defined as a string: https://github.com/okta/okta-sdk-golang/pull/112/files#diff-ad4db84096a7b93a350a17acf8082116R49

I'm guessing this works, but is there any need to specify encoding when converting string, to byte[]? Is it assumed to be UTF-8?

File: okta/requestExecutor.go [hunk 4]
```diff
@@ -74,7 +95,83 @@ func (re *RequestExecutor) NewRequest(method string, url string, body interface{
...
+			if err != nil {
+				return nil, err
+			}
+
+			var tokenRequestBuff io.ReadWriter
+			tokenRequestUrl := re.config.Okta.Client.OrgUrl + "/oauth2/v1/token"
+			tokenRequestUrl += "?grant_type=client_credentials"
+			tokenRequestUrl += "&scope=" + strings.Join(re.config.Okta.Client.Scopes, " ")
+			tokenRequestUrl += "&client_assertion_type=urn:ietf:params:oauth:client-assertion-type:jwt-bearer"
+			tokenRequestUrl += "&client_assertion=" + clientAssertion
```
Comment (aarongranick-okta) [inline okta/requestExecutor.go]: Should these URI parameters be URL-encoded?

File: okta/requestExecutor.go (hunk 3 above)
Comment (bretterer) [reply okta/requestExecutor.go]: Nope.   all good here

File: README.md [hunk 5]
```diff
@@ -377,6 +394,32 @@ okta:
     token: {apiToken}
 ```
 
+When you use OAuth 2.0 the full YAML configuration looks like:
+
+```yaml
+okta:
+  client:
+    connectionTimeout: 30 # seconds
+    oktaDomain: "https://{yourOktaDomain}"
```
Comment (aarongranick-okta) [inline README.md]: should this be `orgUrl` ?

File: okta/config.go (hunk 2 above)
Comment (bretterer) [reply okta/config.go]: Fixed

File: README.md (hunk 1 above)
Comment (bretterer) [reply README.md]: Fixed

File: README.md (hunk 5 above)
Comment (bretterer) [reply README.md]: fixed

**Comment by bretterer** [comment]
> * It is requiring my oktaOrg to begin with https://. Is this consistent with other SDKs? It makes sense from an end-user perspective but local rain is typically on http://

Yes, this is common.  There is a config for disabling HTTPS checks you can use for testing purposes when local.

**Pull request merged**